packages/dd-trace/src/llmobs/sdk.js

```diff
@@ -0,0 +1,448 @@
+    const name = getName(kind, options)
+
+    const {
+      spanOptions,
+      ...llmobsOptions
+    } = this._extractOptions(options)
+
+    if (fn.length > 1) {
+      return this._tracer.trace(name, spanOptions, (span, cb) => {
+        const oldStore = storage.getStore()
+        const parentLLMObsSpan = oldStore?.llmobsSpan
+        if (this.enabled) storage.enterWith({ ...oldStore, llmobsSpan: span })
+
+        this._tagger.setLLMObsSpanTags(span, valid && kind, {
+          ...llmobsOptions,
+          parentLLMObsSpan
+        })
+
+        return fn(span, err => {
+          if (this.enabled) storage.enterWith(oldStore)
+          cb(err)
+        })
+      })
+    }
+
+    return this._tracer.trace(name, spanOptions, span => {
+      const oldStore = storage.getStore()
+      const parentLLMObsSpan = oldStore?.llmobsSpan
+      if (this.enabled) storage.enterWith({ ...oldStore, llmobsSpan: span })
+
+      this._tagger.setLLMObsSpanTags(span, valid && kind, {
+        ...llmobsOptions,
+        parentLLMObsSpan
+      })
+
+      try {
+        const result = fn(span)
+
+        if (result && typeof result.then === 'function') {
+          return result.then(value => {
+            if (this.enabled) storage.enterWith(oldStore)
+            return value
+          }).catch(err => {
+            if (this.enabled) storage.enterWith(oldStore)
+            throw err
+          })
+        }
+
+        if (this.enabled) storage.enterWith(oldStore)
+        return result
+      } catch (e) {
+        if (this.enabled) storage.enterWith(oldStore)
+        throw e
+      }
+    })
+  }
+
+  wrap (options = {}, fn) {
+    if (typeof options === 'function') {
+      fn = options
+      options = {}
+    }
+
+    const kind = options.kind
+    const name = getName(kind, options, fn)
+
+    const {
+      spanOptions,
+      ...llmobsOptions
+    } = this._extractOptions(options)
+
+    const llmobs = this
+
+    function wrapped () {
+      if (!llmobs.enabled) {
+        logger.warn('Span started while LLMObs is disabled. Spans will not be sent to LLM Observability.')
+      }
+
+      const valid = validKind(kind)
+      if (!valid) {
+        logger.warn(`Invalid span kind specified: ${kind}. Span will not be sent to LLM Observability.`)
+      }
+
+      const span = llmobs._tracer.scope().active()
+      const oldStore = storage.getStore()
+      const parentLLMObsSpan = oldStore?.llmobsSpan
+      if (llmobs.enabled) storage.enterWith({ ...oldStore, llmobsSpan: span })
+
+      llmobs._tagger.setLLMObsSpanTags(span, valid && kind, {
+        ...llmobsOptions,
+        parentLLMObsSpan
+      })
+      llmobs.annotate(span, { inputData: getFunctionArguments(fn, arguments) })
+
+      try {
+        const result = fn.apply(this, arguments)
+
+        if (result && typeof result.then === 'function') {
+          return result.then(value => {
+            if (value && kind !== 'retrieval' && !LLMObsTagger.tagMap.get(span)?.[OUTPUT_VALUE]) {
```

Comment:
will this still attempt to auto annotate the output for an llm-kind span? Asking because LLM span outputs are in message format so we should just avoid auto-annotating here entirely.

Reply:
ahhh yeah, i actually missed that our `_model_decorator`s do not auto annotate at all. i'll add a check here for `llm`, and then input annotation check for `llm` and `embedding`, and just write a couple small regression tests!